### ISSUE TITLE
Support read_only and super_read_only global variables

### ISSUE THREAD

**Issue by kolbe** [pingcap/tidb#18426]
## Feature Request

**Is your feature request related to a problem? Please describe:**

Today, there's no easy way to make a cluster read-only. This could be useful in migration situations, and in situations where TiDB-Binlog or TiCDC is used to sync an upstream TiDB cluster to a downstream TiDB cluster (that may require additional support in the privilege system).

**Describe the feature you'd like:**

I'd like the read_only system variable to work basically the way it works in MySQL (https://dev.mysql.com/doc/refman/5.7/en/server-system-variables.html#sysvar_read_only). That means only users with the SUPER privilege can execute statements that update table data.

Support for super_read_only should also be added.

To support the TiCDC and TiDB-Binlog use cases, a user account identified as a "replication secondary" should be allowed to write to a cluster with either read_only or super_read_only.

We should consider how the action of actually setting read_only/super_read_only on an *upstream* cluster impacts a downstream cluster.

**Describe alternatives you've considered:**

There are many ways to prevent users from *connecting* to TiDB, but it's difficult to emulate the real behavior of read_only that allows users to connect but not to modify data.

**Teachability, Documentation, Adoption, Migration Strategy:**

Useful for maintenance, migration, replication.

**Comment by zz-jason** [comment]
@IANTHEREAL Could you help us to clarify the use case and value of this feature request?

**Comment by zz-jason** [comment]
@kolbe `read-only` sys variable can block all the write operations on the whole cluster. Would it help to only make certain tables read-only?

**Comment by IANTHEREAL** [comment]
Roughly there are the following scenarios

- Disaster Recovery Cluster Replication - can set read-only at the cluster level.

- cyclical replication - because each cluster has data write, read-only cluster not applicable

- Other replication scenarios - not sure if there are data write, more reasonable is to control read-only granularity. For example, making some tables read-only, it requires a careful design of table-level read-only syntax

**Comment by IANTHEREAL** [comment]
in addition, I think the read-only function may be necessary for some financial replication scenarios. 

https://github.com/pingcap/ticdc/issues/691 may need it

right? @amyangfei

**Comment by amyangfei** [comment]
> in addition, I think the read-only function may be necessary for some financial replication scenarios.
> 
> [pingcap/ticdc#691](https://github.com/pingcap/ticdc/issues/691) may need it
> 
> right? @amyangfei

read-only mode will help in the primary-secondary switch scenario

Assuming the topology as follows
```
outer write traffic -> cluster A -> TiCDC -> cluster B
```
Write traffic goes into cluster A and we use TiCDC to replicate data from A to B in real-time. If we want to switch write traffic from A to B, we need to guarantee the data consistency between A and B when switching.

**Comment by IANTHEREAL** [comment]
I will explain the relationship between read-only and consistent replication.

Suppose there are a replication scenario: cluster A -> ticdc -> cluster B, and make sure that cluster B is always in the transaction (some transaction ts of cluster A) consistent with the state.

One solution is to save the mapping relationship in the cluster B:  ticdc checkpoint ts <-> consistent transaction ts, and meet two conditions:

1. in cluster A,  all transactions whose transaction ts <= ticdc checkpoint ts have been replicated to the cluster B, and the corresponding transaction in cluster B hash transaction ts <= consistent transaction ts.
2. in cluster A, the transactions whose transaction ts > ticdc checkpoint and that has been replicated to cluster B, which the corresponding transaction in cluster B hash transaction ts > consistent transaction ts.

**Comment by kolbe** [comment]
@zz-jason making specific tables read-only does not help with the use case I'm interested in here. This FR relates only to the use case where a cluster denies writes except by users with specific global privileges ("SUPER" or "replication writer")

**Comment by scsldb** [comment]
@kolbe  @zz-jason  **Notes**:  We accept `read-only global variables` requirements, but `super-read-only global variables` requirements will be adjusted according to the actual situation late.

**Comment by zz-jason** [comment]
Yeah, after discussion `read-only` is a high priority issue. We'll work on the technical design firstly.

**Comment by zz-jason** [comment]
I'm going to change the title of this issue to "support `read-only` global variable". Let's discuss the `super-read-only` variable in another feature request issue.

**Comment by nullnotnil** [comment]
See also: https://github.com/pingcap/tidb/issues/9172 (MySQL can do both a read-only and an "Offline Mode"). `read-only` sounds like higher priority though.

**Comment by SunRunAway** [comment]
@ichn-hu PTAL

**Comment by morgo** [comment]
**Note:** TiDB does not yet support `read_only` mode, but it does support `tidb_restricted_read_only`. It's a little bit different, see the design doc here: https://github.com/pingcap/tidb/blob/master/docs/design/2021-06-23-restricted-read-only.md
